### ISSUE TITLE
Hiding constants from closure_convert and the AD system

### ISSUE THREAD

**Issue by ahoenselaar** [google/jax#6415]
Use case: Higher order functions (like ODE solvers and optimizers) with custom VJP rules.

Consider this example form the JAX documentation for `closure_convert`:

```
def minimize(objective_fn, x0):
  converted_fn, aux_args = closure_convert(objective_fn, x0)
  return _minimize(converted_fn, x0, *aux_args)

@partial(custom_vjp, nondiff_argnums=(0,))
def _minimize(objective_fn, x0, *args):
  z = objective_fn(x0, *args)
  # ... find minimizer x_opt ...
  return x_opt

def fwd(objective_fn, x0, *args):
  y = _minimize(objective_fn, x0, *args)
  return y, (y, args)

def rev(objective_fn, res, g):
  y, args = res
  y_bar = g
  # ... custom reverse-mode AD ...
  return x0_bar, *args_bars

_minimize.defvjp(fwd, rev)
```

Imagine now that `objective_fn` depends on a large matrix of constants `c`,  e.g. captured via closure. Then `closure_convert` pulls this matrix out and makes it an explicit argument of `converted_fn`.
During the reverse pass, `c_bar` has to be calculated as part of `args_bar`. This computation can be expensive and complex but it is not needed if differentiation with respect to `c` is never required or does not make sense.
There should be a way for the author of `objective_fn` to annotate `c` accordingly. This cannot be done at the level of the minimizer, where it is not known which `args` are needed in the VJP and which ones are not.

In simple cases, it is conceivable that XLA manages to optimize out the unused `c_bar`. That does not work for more complex cases, e.g. [the backward pass of an ODE integrator that has to integrate an augmented system of ODEs](https://github.com/google/jax/blob/3ac809ede30d961b873ca62f9828090709ced6be/jax/experimental/ode.py#L221).

In our concrete application (ODE integration where the dynamics function captures a large matrix), the impact of calculating `c_bar` is such that XLA consumes >44 GB of RAM during the compilation stage while attempting to generate code for the Jacobian of the dynamics function with respect to each and every element of `c`.

Can `closure_convert` be modified to not extract certain captured variables? At the moment, it ignores integer arrays, for example.
I also see parallels to https://github.com/google/jax/issues/5913, which is more focused on the JVP case.

**Comment by froystig** [comment]
The `custom_vjp` convention is to return `None` for a symbolic zero cotangent. See [the example here](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html#gradient-clipping). This would amount to setting the `c_bar` that you describe to `None` instead of filling a large matrix.

In principle, this would still define a cotangent (at very low cost), even in cases where there ought to be no meaningful derivative at all (zero or otherwise). Still, would this suffice as a workaround in your case?

Even if so, possible improvements from our end:
1. Mention the `None` convention in the `custom_vjp` docstring directly.
2. Allow for custom derivative arguments that do not participate in differentiation.

**Comment by ahoenselaar** [comment]
The custom VJP is defined on `_minimize`. The author of `_minimize` or some other higher-order function in a library cannot determine for which elements in `args` to return `None`. As `closure_convert` has to be used for correctness, the forward pass is handed a fairly opaque list of `args`.

Only the author of `objective_fn`, i.e. the user of the library, can determine which arguments or captured values should be ignored.

**Comment by ahoenselaar** [comment]
For comparison: [PyTorch autograd allows tensors to be excluded from the autograd DAG](https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html#exclusion-from-the-dag).

Arguably, JAX autograd automatically opts in every floating point tensor. For the scenario described here, I see two solutions:

- An opt-out facility for the matrix of constants `c`.
- I believe that tracing should reveal that `c_bar` is unused. If that information could be made available to the backward pass, `None` could be returned for the appropriate arguments.

**Comment by shoyer** [comment]
I wonder if it would help to dynamically create the `custom_vjp` decorated function _inside_ `minimize`? This might even be reasonably maintainable, as long as the library provides suitable helper functions for implementing the gradient rules.

**Comment by shoyer** [comment]
> For comparison: [PyTorch autograd allows tensors to be excluded from the autograd DAG](https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html#exclusion-from-the-dag).
> 
> Arguably, JAX autograd automatically opts in every floating point tensor. For the scenario described here, I see two solutions:
> 
> * An opt-out facility for the matrix of constants `c`.
> * I believe that tracing should reveal that `c_bar` is unused. If that information could be made available to the backward pass, `None` could be returned for the appropriate arguments.

JAX does have functionality like this, via `lax.stop_gradient`.

The problem is that it only works for normal JAX autodiff rules, i.e., with VJP implemented via transpose of JVP. This currently requires the lower-level JAX primitive API.

**Comment by froystig** [comment]
If possible, let's separate:
1. concerns about AD efficiency (unwittingly computing a cotangent that is never read downstream); from
2. concerns about AD correctness (a cotangent is mathematically ill-defined, but jax still requires a definition); from
3. limitations of the `closure_convert` utility.

Is this request captured entirely by any one of the above? A complete example would help.

Based on more recent comments, my new guess is that this is scoped to item 3. Today's `closure_convert` extracts all floating-point closure-captured arrays as a heuristic. If `closure_convert` hoisted only arguments that could possibly participate in differentiation instead, this would reduce the concerns of items 1 and 2 when using `closure_convert` together with `custom_vjp`.  Does that sound right?

If so, then that's what [this TODO comment](https://github.com/google/jax/blob/ad342419b87fbe08afe9a5f959704dee3a654a0d/jax/custom_derivatives.py#L825-L827) from @mattjj and me is about, and we can use this issue to track it.

**Comment by ahoenselaar** [comment]
From my point of view:
The need to use `closure_convert` and its limitations (3) are negatively affecting AD efficiency (1).

**Comment by froystig** [comment]
OK, let's track that here then. I suspect it will be easiest to tackle `closure_convert`'s heuristic first. If not, we can consider our options for changing the calling convention of a `custom_jvp`'s forward and reverse steps, or maybe improving DCE in AD somehow.

By the way, #5913 is about the JVP analogue of (2).

**Comment by ahoenselaar** [comment]
Have any paths forward to modify `closure_convert` been identified?

**Comment by froystig** [comment]
#7369 improves the heuristic within `closure_convert` to hoist fewer values. I can't verify the fix without a reproducing example. @ahoenselaar: with #7369 merged now, is your large constant matrix still hoisted or do we successfully avoid it?

**Comment by froystig** [comment]
Closing because I expect that the issue is addressed based on its description in this thread. Please reopen if it is not!
